src/ui/marker.js

```diff
@@ -444,35 +448,56 @@ export default class Marker extends Evented {
         return this;
     }
 
-    _updateOcclusion() {
-        if (!this._occlusionTimer) {
-            this._occlusionTimer = setTimeout(this._onOcclusionTimer.bind(this), 60);
-        }
-    }
+    _evaluateOpacity(requestAnimationFrame: boolean = false) {
+        const position = this._pos ? this._pos.sub(this._transformedOffset()) : null;
 
-    _clearOcclusionTimer() {
-        if (this._occlusionTimer) {
-            clearTimeout(this._occlusionTimer);
-            this._occlusionTimer = null;
+        if (!this._withinScreenBounds(position)) {
+            this._fadeTimer = null;
+            return;
         }
-    }
 
-    _onOcclusionTimer() {
-        const tr = this._map.transform;
-        const pos = this._pos ? this._pos.sub(this._transformedOffset()) : null;
-        if (pos && pos.x >= 0 && pos.x < tr.width && pos.y >= 0 && pos.y < tr.height) {
-            // calculate if occluded.
-            const raycastLoc = this._map.unproject(pos);
+        const mapLocation = this._map.unproject(position);
+
+        let terrainOccluded = false;
+        if (this._map.getTerrain()) {
             const camera = this._map.getFreeCameraOptions();
             if (camera.position) {
                 const cameraPos = camera.position.toLngLat();
-                const raycastDistance = cameraPos.distanceTo(raycastLoc);
-                const posDistance = cameraPos.distanceTo(this._lngLat);
-                const occluded = raycastDistance < posDistance * 0.9;
-                this._element.classList.toggle('mapboxgl-marker-occluded', occluded);
+                // the distance to the marker lat/lng + marker offset location
+                const offsetDistance = cameraPos.distanceTo(mapLocation);
+                const distance = cameraPos.distanceTo(this._lngLat);
+                terrainOccluded = offsetDistance < distance * 0.9;
             }
         }
-        this._occlusionTimer = null;
+
+        const fogOpacity = this._map.queryFogOpacity(mapLocation);
+        const opacity = (1.0 - fogOpacity) * (terrainOccluded ? TERRAIN_OCCLUDED_OPACITY : 1.0);
+        const updateStyle = () => {
+            this._element.style.opacity = `${opacity}`;
+            if (this._popup) this._popup._setOpacity(`${opacity}`);
+        };
+
+        if (requestAnimationFrame) {
+            window.requestAnimationFrame(updateStyle);
```

Comment:
What is the purpose of updating this within an animation frame?

When the callback gets called the marker might not exist anymore. I don't think it would error though.

Reply:
Addressed in https://github.com/mapbox/mapbox-gl-js/pull/10564/commits/158dece43a7fcf613e5376c960cd1991c4590a42